core/src/main/scala/org/apache/spark/Dependency.scala

```diff
@@ -95,6 +97,41 @@ class ShuffleDependency[K: ClassTag, V: ClassTag, C: ClassTag](
   val shuffleHandle: ShuffleHandle = _rdd.context.env.shuffleManager.registerShuffle(
     shuffleId, this)
 
+  // By default, shuffle merge is enabled for ShuffleDependency if push based shuffle is enabled
+  private[this] var _shuffleMergeEnabled = Utils.isPushBasedShuffleEnabled(rdd.sparkContext.getConf)
+
+  def setShuffleMergeEnabled(shuffleMergeEnabled: Boolean): Unit = {
+    _shuffleMergeEnabled = shuffleMergeEnabled
+  }
+
+  def shuffleMergeEnabled : Boolean = _shuffleMergeEnabled
+
+  /**
+   * Stores the location of the list of chosen external shuffle services for handling the
+   * shuffle merge requests from mappers in this shuffle map stage.
+   */
+  private[this] var _mergerLocs: Seq[BlockManagerId] = Nil
+
+  /**
+   * Stores the information about whether the shuffle merge is finalized for the shuffle map stage
+   * associated with this shuffle dependency
+   */
+  private[this] var _shuffleMergedFinalized: Boolean = false
```

Comment:
Is it necessary to introduce shuffleMergeFinalized in this PR?

Reply:
On a closer look, We are using `shuffleMergeFinalized` as part of the `prepareShuffleServicesForShuffleMapStage` step but without having the finalized it doesn't make sense to have that. I can wait and get @zhouyejoe changes also included once he posts the PR (https://github.com/linkedin/spark/commit/818eca4199574514a1a7a1bd0b34a1d4da94ff6e).

Reply:
For now have removed the `shuffleMergeFinalized` check eventually we need to add this back.